app/main.py

```diff
@@ -24,22 +45,102 @@ def get_latest_release() -> models.LatestRelease:
+    # Sync latest version tag using GitHub API
     app.latest_release = get_latest_release()
+    yield
+
+
+@repeat_every(seconds=15 * 60)  # every 15 minutes
+def update_version():
+    """Sync latest version tag using GitHub API"""
+    app.latest_release = get_latest_release()
+
+
+# Fields to store per visit
+fieldnames = [
+    "version_multiqc",
+    "version_python",
+    "operating_system",
+    "installation_method",
+    "ci_environment",
+]
+
+# Thread-safe in-memory buffer to accumulate recent visits before writing to the CSV file
+visit_data = []
+visit_data_lock = Lock()
+
+
+@app.get("/version")  # log a visit
+async def version(
+    version_multiqc: str | None = None,
+    version_python: str | None = None,
+    operating_system: str | None = None,
+    installation_method: str | None = None,
+    ci_environment: str | None = None,
+):
+    """
+    Endpoint for MultiQC that returns the latest release, and logs
+    the visit along with basic user environment detail.
+    """
+    with visit_data_lock:
+        visit_data.append(
+            {
+                "timestamp": datetime.datetime.now().isoformat(),
+                "version_multiqc": version_multiqc,
+                "version_python": version_python,
+                "operating_system": operating_system,
+                "installation_method": installation_method,
+                "ci_environment": ci_environment,
+            }
+        )
+    return models.VersionResponse(latest_release=app.latest_release)
+
+
+# Path to a buffer CSV file to persist recent visits before dumping to the database
+# In the same folder as this script
+CSV_FILE_PATH = Path(__file__).parent / "visits.csv"
+
+
+@repeat_every(seconds=10)  # every 10 seconds
+async def persist_visits():
+    """Write in-memory visits to a CSV file"""
+    global visit_data
+    with visit_data_lock:
+        if visit_data:
+            logger.debug(f"Persisting {len(visit_data)} visits to CSV {CSV_FILE_PATH}")
+            with open(CSV_FILE_PATH, mode="a") as file:
+                writer = csv.DictWriter(file, fieldnames=["timestamp"] + fieldnames)
+                for row in visit_data:
+                    writer.writerow(row)
+            visit_data = []
+
+
+@repeat_every(seconds=60 * 60 * 1)  # every hour
+async def summarize_visits():
+    with visit_data_lock:
+        df = pd.read_csv(CSV_FILE_PATH, sep=",", names=["timestamp"] + fieldnames)
+        df["start"] = pd.to_datetime(df["timestamp"])
+        df["end"] = df["start"] + pd.to_timedelta("1min")
+        df["start"] = df["start"].dt.strftime("%Y-%m-%d %H:%M")
+        df["end"] = df["end"].dt.strftime("%Y-%m-%d %H:%M")
+        df = df.drop(columns=["timestamp"])
+        # replace nan with "Unknown"
+        df = df.fillna("Unknown")  # df.groupby will fail if there are NaNs
+        # Summarize visits per user per minute
+        minute_summary = df.groupby(["start", "end"] + fieldnames).size().reset_index(name="count")
```

Comment:
Does this do the entire CSV file every time? See above comment about cleaning that file, will the csv just grow and grow and this code get slower and slower?

Reply:
Yep, forgot to add the code to rotate it